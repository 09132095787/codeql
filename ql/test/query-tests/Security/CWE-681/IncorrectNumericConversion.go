package main

import (
	"math"
	"strconv"
)

func main() {

}

type Something struct {
}
type Config struct {
}
type Registry struct {
}

func LookupTarget(conf *Config, num int32) (int32, error) {
	return 567, nil
}
func LookupNumberByName(reg *Registry, name string) (int32, error) {
	return 567, nil
}
func lab(s string) (*Something, error) {
	num, err := strconv.Atoi(s)

	if err != nil {
		number, err := LookupNumberByName(&Registry{}, s)
		if err != nil {
			return nil, err
		}
		num = int(number)
	}
	target, err := LookupTarget(&Config{}, int32(num))
	if err != nil {
		return nil, err
	}

	// convert the resolved target number back to a string

	s = strconv.Itoa(int(target))

	return nil, nil
}

const CustomMaxInt16 = 1<<15 - 1

type CustomInt int16

// these should be caught:
func upperBoundIsNOTChecked(input string) {
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		_ = int8(parsed)
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		_ = int16(parsed)
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		_ = int32(parsed)
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		_ = uint8(parsed)
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		_ = uint16(parsed)
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		_ = uint32(parsed)
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		_ = float32(parsed)
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		// NOTE: byte is uint8
		_ = byte(parsed)
	}
	{
		// using custom type:
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		_ = CustomInt(parsed)
	}

}

// these should NOT be caught:
func upperBoundIsChecked(input string) {
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		if parsed < math.MaxInt8 {
			_ = int8(parsed)
		}
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		if parsed < math.MaxInt16 {
			_ = int16(parsed)
		}
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		if parsed > 0 {
			_ = int32(parsed)
		}
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		if parsed < math.MaxInt32 {
			_ = int32(parsed)
		}
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		if parsed < math.MaxUint8 {
			_ = uint8(parsed)
		}
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		if parsed < math.MaxUint16 {
			_ = uint16(parsed)
		}
	}
	{
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		if parsed < math.MaxUint8 {
			_ = byte(parsed)
		}
	}
	{ // multiple `and` conditions
		parsed, err := strconv.Atoi(input)
		if err == nil && 1 == 1 && parsed < math.MaxInt8 {
			_ = int8(parsed)
		}
	}
	{ // custom maxInt16
		parsed, err := strconv.Atoi(input)
		if err != nil {
			panic(err)
		}
		if parsed < CustomMaxInt16 {
			_ = int16(parsed)
		}
	}
}
