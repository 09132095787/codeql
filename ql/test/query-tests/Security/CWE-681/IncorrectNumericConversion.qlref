Security/CWE-681/IncorrectNumericConversion.ql