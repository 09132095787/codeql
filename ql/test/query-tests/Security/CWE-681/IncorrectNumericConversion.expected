edges
| IncorrectNumericConversion.go:26:2:26:28 | ... := ...[0] : int | IncorrectNumericConversion.go:35:41:35:50 | type conversion |
| IncorrectNumericConversion.go:54:3:54:36 | ... := ...[0] : int | IncorrectNumericConversion.go:58:7:58:18 | type conversion |
| IncorrectNumericConversion.go:61:3:61:36 | ... := ...[0] : int | IncorrectNumericConversion.go:65:7:65:19 | type conversion |
| IncorrectNumericConversion.go:68:3:68:36 | ... := ...[0] : int | IncorrectNumericConversion.go:72:7:72:19 | type conversion |
| IncorrectNumericConversion.go:75:3:75:36 | ... := ...[0] : int | IncorrectNumericConversion.go:79:7:79:19 | type conversion |
| IncorrectNumericConversion.go:82:3:82:36 | ... := ...[0] : int | IncorrectNumericConversion.go:86:7:86:20 | type conversion |
| IncorrectNumericConversion.go:89:3:89:36 | ... := ...[0] : int | IncorrectNumericConversion.go:93:7:93:20 | type conversion |
| IncorrectNumericConversion.go:96:3:96:36 | ... := ...[0] : int | IncorrectNumericConversion.go:100:7:100:21 | type conversion |
| IncorrectNumericConversion.go:103:3:103:36 | ... := ...[0] : int | IncorrectNumericConversion.go:108:7:108:18 | type conversion |
| IncorrectNumericConversion.go:112:3:112:36 | ... := ...[0] : int | IncorrectNumericConversion.go:116:7:116:23 | type conversion |
nodes
| IncorrectNumericConversion.go:26:2:26:28 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:35:41:35:50 | type conversion | semmle.label | type conversion |
| IncorrectNumericConversion.go:54:3:54:36 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:58:7:58:18 | type conversion | semmle.label | type conversion |
| IncorrectNumericConversion.go:61:3:61:36 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:65:7:65:19 | type conversion | semmle.label | type conversion |
| IncorrectNumericConversion.go:68:3:68:36 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:72:7:72:19 | type conversion | semmle.label | type conversion |
| IncorrectNumericConversion.go:75:3:75:36 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:79:7:79:19 | type conversion | semmle.label | type conversion |
| IncorrectNumericConversion.go:82:3:82:36 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:86:7:86:20 | type conversion | semmle.label | type conversion |
| IncorrectNumericConversion.go:89:3:89:36 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:93:7:93:20 | type conversion | semmle.label | type conversion |
| IncorrectNumericConversion.go:96:3:96:36 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:100:7:100:21 | type conversion | semmle.label | type conversion |
| IncorrectNumericConversion.go:103:3:103:36 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:108:7:108:18 | type conversion | semmle.label | type conversion |
| IncorrectNumericConversion.go:112:3:112:36 | ... := ...[0] : int | semmle.label | ... := ...[0] : int |
| IncorrectNumericConversion.go:116:7:116:23 | type conversion | semmle.label | type conversion |
#select
| IncorrectNumericConversion.go:26:2:26:28 | ... := ...[0] : int | IncorrectNumericConversion.go:26:2:26:28 | ... := ...[0] : int | IncorrectNumericConversion.go:35:41:35:50 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
| IncorrectNumericConversion.go:54:3:54:36 | ... := ...[0] : int | IncorrectNumericConversion.go:54:3:54:36 | ... := ...[0] : int | IncorrectNumericConversion.go:58:7:58:18 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
| IncorrectNumericConversion.go:61:3:61:36 | ... := ...[0] : int | IncorrectNumericConversion.go:61:3:61:36 | ... := ...[0] : int | IncorrectNumericConversion.go:65:7:65:19 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
| IncorrectNumericConversion.go:68:3:68:36 | ... := ...[0] : int | IncorrectNumericConversion.go:68:3:68:36 | ... := ...[0] : int | IncorrectNumericConversion.go:72:7:72:19 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
| IncorrectNumericConversion.go:75:3:75:36 | ... := ...[0] : int | IncorrectNumericConversion.go:75:3:75:36 | ... := ...[0] : int | IncorrectNumericConversion.go:79:7:79:19 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
| IncorrectNumericConversion.go:82:3:82:36 | ... := ...[0] : int | IncorrectNumericConversion.go:82:3:82:36 | ... := ...[0] : int | IncorrectNumericConversion.go:86:7:86:20 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
| IncorrectNumericConversion.go:89:3:89:36 | ... := ...[0] : int | IncorrectNumericConversion.go:89:3:89:36 | ... := ...[0] : int | IncorrectNumericConversion.go:93:7:93:20 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
| IncorrectNumericConversion.go:96:3:96:36 | ... := ...[0] : int | IncorrectNumericConversion.go:96:3:96:36 | ... := ...[0] : int | IncorrectNumericConversion.go:100:7:100:21 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
| IncorrectNumericConversion.go:103:3:103:36 | ... := ...[0] : int | IncorrectNumericConversion.go:103:3:103:36 | ... := ...[0] : int | IncorrectNumericConversion.go:108:7:108:18 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
| IncorrectNumericConversion.go:112:3:112:36 | ... := ...[0] : int | IncorrectNumericConversion.go:112:3:112:36 | ... := ...[0] : int | IncorrectNumericConversion.go:116:7:116:23 | type conversion | Incorrect type conversion of int from strconv.Atoi result to another numeric type |
